### PR TITLE
Retain all source IDs on VariantContext merge + test

### PR DESCRIPTION
**Pull request by lbergelson**
@bbimber I've added a test to your pr https://github.com/broadinstitute/gatk/pull/8750.  Should be good to merge if tests pass.

### CODE REVIEW

**Review by bbimber — APPROVED** [review]
@lbergelson great, thanks for adding the test

**Review by bbimber — APPROVED** [review]

**Review by bbimber — APPROVED** [review]

**Comment by bbimber** [comment]
@lbergelson: this might be permission-related, but despite entering an approving review github is asking for another. I assume that's b/c I dont have write permissions in this project?

**Comment by lbergelson** [comment]
@bbimber Hmn, yeah,  think it needs someone who has direct write access.  I'll get a thumb from a teammate.  Thanks for looking at it and for the pr!

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java
Comment (lbergelson) [reply src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java:1221]: It's a good question.  Apparently this is what GATK3 did.  @bbimber Any objection to capping it at some reasonable length?

Comment (bbimber) [reply src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java:1221]: Honestly, I dont have personal use-cases around this, but on DISCVR-seq (which ported GATK3's CombineVariants) I get a surprising amount of requests from people around the lack-of or changes to this feature. 

I have no real objection to some kind of cap on length. What length would you find reasonable? I suppose that value could be parameterized, but this would require exposing an argument to simpleMerge() and maybe overloading that method.

Anyway, this feels like a valid but fringe case to me. I dont have strong opinions one way or another.

Comment (droazen) [reply src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java:1221]: Combining hundreds or thousands of VCF records is something we actually do frequently in practice. I think a parameterized option would be overkill, but some kind of sensible hardcoded max length for the source field seems appropriate.

Comment (droazen) [reply src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java:1221]: @lbergelson I defer to you to do something reasonable in this PR to address the effects of this change on VCF file size for large cohorts.

Comment (bbimber) [reply src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java:1221]: FWIW, after thinking O/N on this I would probably argue for overloading the method with one that adds arguments for "storeAllVcfSources", and "maxSourceLength", but keep the existing signature where storeAllVcfSources=False, and maxSourceLength=-1 (which is moot since the existing behavior of taking the first source would remian). this preserves the status quo, but callers could opt-in to GATK3-like behavior. I am happy to make those changes.

**Comment by bbimber** [comment]
Hi @lbergelson and @droazen: it looks like @lbergelson added a limit to avoid excessive sources. Does that satisfy issues remaining on this PR?

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java:1225]: I am concerned about doing a sort on the source names for every single merged VCF record we're writing -- with thousands or tens of thousands of samples, that would get expensive!

**Review by droazen — CHANGES_REQUESTED** [review]
@lbergelson Back to you with another concern

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java
Comment (bbimber) [reply src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java:1225]: That's reasonable - what about what I proposed a little earlier in this thread and overloading the method to make the existing default case unchanged (list the first source), but have the ability for code to opt-in to the behavior of tracking all sources? I am happy to write that.

Comment (droazen) [reply src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java:1225]: @bbimber That option is looking more attractive -- if you'd like to implement it, please go ahead and tag us for review. If we don't touch the default behavior this will be much easier to get in!

**Comment by bbimber** [comment]
Hey @lbergelson, @droazen or @jamesemery: I'm hoping we could finalize this issue. I am unable to commit directly to this branch. To respond to the suggestions from @droazen I made this PR which i am hoping someone can merge into this feature branch (https://github.com/broadinstitute/gatk/pull/8871). 

To recap:
- This PR is driven by a request from some of DISCVRseq's users to restore a GATK3 behavior where all source IDs are saved for variants when multiple VCFs are merged.
- This PR would expose this as an opt-in feature, and should not change any default GATK4 behavior. Therefore existing variant merge code should be unchanged.

**Comment by bbimber** [comment]
Hey @lbergelson, @droazen or @jamesemery: sorry to keep bugging you all here, but I'm not sure what else to do here. I'm hoping we can finalize and close out this feature, or identify what is needed to do so. I recapped the state of this PR in my post above. 

I could make a clean feature branch with everything in one place if you want; however, if someone with write permissions could simply approve (https://github.com/broadinstitute/gatk/pull/8871) , merge into this branch, and kick off tests that is all I need right now.

**Pull request closed without merging**
